### PR TITLE
feat: inject checkpoint summary into session restore to prevent confused agent state (#842)

### LINKED ISSUE #842: feat: inject checkpoint summary into session restore to prevent confused agent state
## Problem

When `ao session restore` revives a crashed session, it recreates the tmux session and relies entirely on the agent's internal session resumption (e.g. `claude --resume`, OpenCode session reuse). There is **no validation** that the agent actually understands its restored state.

A crash mid-task can leave the agent in a half-baked mental state:
- Agent thinks it committed a file, but the commit never happened
- Agent thinks it completed step 3 of 5, but step 3 was only half-done
-  - Staged/unstaged git changes don't match the agent's conversation history
- Half-written files on disk that the agent doesn't re-read

The lifecycle manager's CI feedback loop catches *some* of this reactively (broken code → CI fail → route errors back), but there's no **proactive** validation at restore time.

## Proposed Solution

Inject a **checkpoint summary** into the agent's context immediately after restore, giving it ground truth about the actual workspace state.

### 1. Periodic checkpoint writing (lifecycle manager)

During each poll cycle, for sessions with `status=working`, snapshot the git state:

### 2. Post-restore injection (session manager)

In `restore()`, after the agent boots:
- Read `git status --porcelain`, `git log --oneline -3`, `git diff --stat`
- Read the checkpoint file if it exists
- Compose a summary message and send it to the agent:

### 3. New module: `packages/core/src/checkpoint.ts
- `writeCheckpoint(session, sessionsDir)` — called by lifecycle manager
- `buildCheckpointSummary(session, workspacePath)` — called by restore

## Files to Change

- `packages/core/src/session-manager.ts` — `restore()`: inject checkpoint after agent boots
- `packages/core/src/lifecycle-manager.ts` — `pollAll()`: write periodic checkpoints
- `packages/core/src/prompt-builder.ts` — optionally add checkpoint awareness to base prompt
- **New:** `packages/core/src/checkpoint.ts` — checkpoint read/write logic

## Why This Matters

Without this, restore is "optimistic replay" — works great for idle-point crashes, poorly for mid-operation crashes. This turns it into "informed resume" — the agent gets ground truth and can self-correct before continuing.

The existing CI feedback loop remains as a safety net, but this prevents the broken code from being pushed in the first place.

### PR DESCRIPTION
**Pull request by nikhilachale**
## Problem

When `ao session restore` revives a crashed session, it relies entirely on the agent's internal session resumption (e.g. `claude --resume`). There's no validation that the agent's conversation history matches the actual workspace state.

A crash mid-task can leave the agent confused:
- Agent thinks it committed a file, but the commit never happened
- Staged/unstaged git changes don't match the agent's memory
- Half-written files on disk that the agent doesn't re-read
- Agent skips or duplicates steps based on stale conversation history
The CI feedback loop catches some damage reactively, but broken code gets pushed before it's caught.

## Solution

Three-part checkpoint system that gives agents ground truth on restore:

### 1. Periodic checkpoints (`checkpoint.ts` — new module)
- `writeCheckpoint()` snapshots git state (commits, staged/modified/untracked files, branch)
- Called by the lifecycle manager every poll cycle for `status=working` sessions
- Atomic writes (tmp → rename) prevent half-written checkpoint files
- Stored in `sessionsDir` alongside metadata — survives worktree deletion

### 2. Post-restore injection (`session-manager.ts`)
- After `restore()` boots the agent, builds a ground-truth summary:
  - Live `git log`, `git status`, `git diff --stat`
  - Last saved checkpoint for comparison
  - Clear instructions to verify state before continuing
- Waits for agent readiness, then sends via `runtime.sendMessage()`
- Non-fatal — agent still works if injection fails

### 3. Agent prompt awareness (`prompt-builder.ts`)
- New "Session Restore" section in `BASE_AGENT_PROMPT`
- Tells agents to trust git state over their memory on restore
- Instructs agents to check for half-written files before editing

## Files Changed

| File | Change |
|------|--------|
| `packages/core/src/checkpoint.ts` | **New** — `writeCheckpoint()` + `buildCheckpointSummary()` |
| `packages/core/src/lifecycle-manager.ts` | Import + call `writeCheckpoint()` for working sessions |
| `packages/core/src/session-manager.ts` | Import + inject checkpoint summary in `restore()` |
| `packages/core/src/prompt-builder.ts` | Add "Session Restore" section to base agent prompt |
| `packages/core/src/__tests__/checkpoint.test.ts` | **New** — 9 test cases covering write, read, summary, edge cases |

## Design Decisions

- **Best-effort, never fatal** — checkpoint failures are silently swallowed. A broken checkpoint must never crash the poll loop or block restore.
- **No external dependencies** — `checkpoint.ts` implements its own atomic write instead of importing `atomicWriteFileSync`, keeping it self-contained.
- **Minimal schema validation** — `readCheckpoint()` checks 3 key fields instead of full Zod validation. Corrupted files gracefully degrade to "no checkpoint found."
- **Checkpoint stored in sessionsDir, not worktree** — survives worktree deletion and recreation during restore.

## Test Results
9/9 passing:
- `writeCheckpoint`: valid JSON, detects untracked/staged files, skips missing workspace
- `buildCheckpointSummary`: includes git state, includes saved checkpoint, handles missing/corrupted checkpoints, returns null for missing workspace

Closes #842

@suraj-markup @AgentWrapper

### CODE REVIEW

File: packages/core/src/checkpoint.ts [hunk 1]
```diff
@@ -0,0 +1,308 @@
...
+  try {
+    const raw = readFileSync(path, "utf-8");
+    const parsed = JSON.parse(raw) as CheckpointData;
+
+    // Minimal schema check — if key fields are missing, treat as corrupted
+    if (!parsed.sessionId || !parsed.timestamp || !parsed.lastCommitHash) {
+      return null;
+    }
+
+    return parsed;
```
[the remark this answers is not in the document]
Comment (nikhilachale) [reply packages/core/src/checkpoint.ts]: This is already handled — `readCheckpoint()` normalizes all array fields via `stringArrayField()` (lines 184-186), which returns `[]` for undefined/non-array values. The test at line 119 (`handles checkpoint missing array fields`) validates this path. No crash is possible here.

File: packages/core/src/checkpoint.ts [hunk 2]
```diff
@@ -0,0 +1,308 @@
...
+    renameSync(tmp, filePath); // atomic on POSIX — single syscall
+  } catch {
+    // Fallback: direct write (non-atomic but better than nothing)
+    try {
+      writeFileSync(filePath, json, "utf-8");
+    } catch {
+      // Silently swallow — checkpoints are best-effort, never crash the caller
+    }
+  }
+}
```
[the remark this answers is not in the document]
Comment (nikhilachale) [reply packages/core/src/checkpoint.ts:153]: Acknowledged. The current code already uses `atomicWriteFileSync` from `atomic-write.ts` (imported on line 19) rather than reimplementing. The `safeWriteJson` wrapper just adds the fallback-to-direct-write behavior and directory creation.

**Comment by suraj-markup** [comment]
Did you discussed this with someone else? 

Also I would first ask what are the scenarios when session will restart without `session restore` because this will help us to fix the root cause.

Means I would look for things that first caused this issue instead of just giving solution for this.

Can you look into this?
